hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/api/TypeMatcher.kt

```diff
@@ -0,0 +1,17 @@
+package uk.gov.justice.digital.hmpps.prisonersearch.search.services.attributesearch.api
+
+import kotlin.reflect.KClass
+import kotlin.reflect.KType
+
+sealed interface TypeMatcher<S> {
+  val attribute: String
+  fun validate() {}
+  fun genericType(): KClass<*> = this::class.genericType()
+  companion object {
+    @JvmStatic
+    fun getSupportedTypes() = TypeMatcher::class.sealedSubclasses.map { it.genericType() }
+  }
+}
+
+private fun KClass<*>.genericType() = this.supertypes.first().genericType()
+internal fun KType.genericType() = this.arguments.first().type!!.classifier as KClass<*>
```

Comment:
This is a hack to workaround the fact that in Kotlin you can't have a reified generic type.